src/library/scala/Array.scala

```diff
@@ -61,10 +61,78 @@ object Array extends FallbackArrayBuilding {
   val emptyShortArray   = new Array[Short](0)
   val emptyObjectArray  = new Array[Object](0)
 
-  implicit def canBuildFrom[T](implicit t: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] =
+  implicit def canBuildFrom[T](implicit tag: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] = {
+    val cls = tag.runtimeClass
+    (if (cls.isPrimitive) {
+      cls match {
```

Comment:
This gets compiled to a cascade of ifs, so order matters. I'd put frequently used primitive types (probably Int, Long, Double, the same ones for which streams are optimized) first.

Reply:
changed to match def make in the same class